numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1464,8 +1430,114 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
                 n_ops.greater_equal);
         break;
     default:
-        result = Py_NotImplemented;
-        Py_INCREF(result);
+        Py_INCREF(Py_NotImplemented);
+        return Py_NotImplemented;
+    }
+    if (result == NULL) {
+        /*
+         * Some failures in the comparison should not be passed on.
+         * In particular, the ufuncs do not have loops for flexible dtype,
+         * so those should be treated separately.  Furthermore, for EQ and NE,
+         * we should never fail.
+         *
+         * Our ideal behaviour would be:
+         *
+         * 1. For EQ and NE:
+         *   - If self and other are scalars, return NotImplemented,
+         *     so that python can assign True of False as appropriate.
+         *   - If either is an array, return an array of bool;
+         *
+         * 2. For LT, LE, GE, GT:
+         *   - If self or other was flexible, return NotImplemented
+         *     (as is in fact the case), so python can raise a TypeError.
+         *   - If other is not convertible, pass on the error (MHvK: not
+         *     sure this is completely logical, but it is what we have).
+         *
+         * However, for backwards compatibilty, we cannot yet return arrays,
+         * so we raise warnings instead.  Furthermore, we warn on python2
+         * for LT, LE, GE, GT, since fall-back behaviour is poorly defined.
+         */
+        PyObject *exc, *val, *tb;
+        int other_is_flexible = 0, ndim_other = 0;
+
+        PyErr_Fetch(&exc, &val, &tb);
+        /*
+         * Determine whether other has a flexible dtype; here, inconvertible
+         * is counted as inflexible.  (This repeats work done in the ufunc,
+         * but OK to waste some time in an unlikely path.)
+         */
+        array_other = (PyArrayObject *)PyArray_FROM_O(other);
+        if (array_other) {
+            other_is_flexible = PyTypeNum_ISFLEXIBLE(PyArray_DESCR(array_other)->type_num);
+            ndim_other = PyArray_NDIM(array_other);
+            Py_DECREF(array_other);
+        }
+        else {
+            PyErr_Clear(); /* we restore the original error if needed */
```

Comment:
Would be clearer if `other_is_flexible` and `ndim_other` were assigned in this path, rather than initialized.